### PR TITLE
UCT/IB: fix CQ credit counting on error handling

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
- fixes rc (verbs/mlx5) and dc (verbs/mlx5)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4107/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4111/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4113/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4120/ for details (Mellanox internal link).

File: src/uct/ib/rc/base/rc_ep.h [hunk 1]
```diff
@@ -353,7 +351,8 @@ uct_rc_txqp_completion_inl_resp(uct_rc_txqp_t *txqp, const void *resp, uint16_t
 static UCS_F_ALWAYS_INLINE uint8_t
 uct_rc_iface_tx_moderation(uct_rc_iface_t *iface, uct_rc_txqp_t *txqp, uint8_t flag)
 {
-    return (txqp->unsignaled >= iface->config.tx_moderation) ? flag : 0;
+    return ((txqp->unsignaled >= iface->config.tx_moderation) ||
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: how about this optimization:
add variable like iface->tx.real_moderation, and make it so iface->tx.real_moderation = min(iface->config.tx_moderation, iface->tx.cq_available)
then, would add overhead only when CQE is used, which is ~1/64 of the times. instead of adding overhead for every send

File: src/uct/ib/rc/base/rc_ep.h [hunk 2]
```diff
@@ -365,7 +364,9 @@ uct_rc_txqp_posted(uct_rc_txqp_t *txqp, uct_rc_iface_t *iface, uint16_t res_coun
         --iface->tx.cq_available;
         UCS_STATS_UPDATE_COUNTER(txqp->stats, UCT_RC_TXQP_STAT_SIGNAL, 1);
     } else {
-        txqp->unsignaled++;
+        ++txqp->unsignaled;
+        /* reserve to error */
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: /* reserve cq credits for every posted operation, in case it would complete with error */

File: src/uct/ib/rc/base/rc_ep.h [hunk 3]
```diff
@@ -365,7 +364,9 @@ uct_rc_txqp_posted(uct_rc_txqp_t *txqp, uct_rc_iface_t *iface, uint16_t res_coun
...
+        --iface->tx.cq_available;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: optimization: can we do something like iface->tx.cq_available -= txqp->unsignaled, only when posting a signaled send?

File: src/uct/ib/rc/verbs/rc_verbs_iface.c [hunk 4]
```diff
@@ -113,11 +114,11 @@ uct_rc_verbs_iface_poll_tx(uct_rc_verbs_iface_t *iface)
             continue;
         }
         uct_rc_verbs_txqp_completed(&ep->super.txqp, &ep->txcnt, count);
+        iface->super.tx.cq_available += count;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c]: why is this change needed?

File: test/gtest/uct/test_peer_failure.cc [hunk 5]
```diff
@@ -43,28 +49,38 @@ class test_uct_peer_failure : public uct_test {
...
     {
         check_caps(UCT_IFACE_FLAG_AM_CB_SYNC);
-        uct_iface_set_am_handler(m_receiver1->iface(), 0, am_dummy_handler,
-                                 reinterpret_cast<void*>(this), UCT_AM_CB_FLAG_SYNC);
-        uct_iface_set_am_handler(m_receiver2->iface(), 0, am_dummy_handler,
-                                 reinterpret_cast<void*>(this), UCT_AM_CB_FLAG_SYNC);
+
+        std::queue<entity *> tmp;
+        entity *e;
+        while (!m_receivers.empty()) {
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc]: pls use "for" loop to iterate over STL container
http://www.cprogramming.com/tutorial/stl/iterators.html

File: test/gtest/uct/test_peer_failure.cc [hunk 6]
```diff
@@ -85,31 +101,47 @@ class test_uct_peer_failure : public uct_test {
...
+               (is_time_out ? UCS_ERR_TIMED_OUT : UCS_ERR_OUT_OF_RANGE);
     }
 
 protected:
-    entity *m_sender, *m_receiver1, *m_receiver2;
-    size_t m_err_count;
-    size_t m_am_count;
-    static size_t m_req_count;
+    entity               *m_sender;
+    std::queue<entity *> m_receivers;
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc]: can use std::vector and then no need to also keep m_nreceivers

File: test/gtest/uct/test_peer_failure.cc [hunk 7]
```diff
@@ -314,10 +343,117 @@ UCS_TEST_P(test_uct_peer_failure_cb, desproy_ep_cb)
...
+        tx_queue_len = 256;
+        UCS_TEST_MESSAGE << name << " setting not found, "
+                         << "taken test default value: " << tx_queue_len;
+    }
+
+    if (GetParam()->tl_name == "rc_mlx5" || GetParam()->tl_name == "dc_mlx5") {
+        UCS_TEST_SKIP_R(GetParam()->tl_name + " temporary disabled");
+    }
+
+#if defined(__arm__) || defined(__aarch64__)
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc]: this is a very specific limitation and would like to avoid it in tests.
Is it possible to reproduce the failure when not all receivers exist in same time?
e.g every time connect to new receiver and fail it.

File: test/gtest/uct/test_peer_failure.cc [hunk 8]
```diff
@@ -314,10 +343,117 @@ UCS_TEST_P(test_uct_peer_failure_cb, desproy_ep_cb)
...
+#endif
+
+    test_uct_peer_failure::m_tx_window  = tx_queue_len / 3;
+
+    test_uct_peer_failure::init();
+
+    check_caps(UCT_IFACE_FLAG_AM_SHORT | UCT_IFACE_FLAG_PENDING);
+}
+
+std::string test_uct_peer_failure_multiple::txqlen_param_name() const
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc]: i would encapsulate this as "size_t get_tx_queue_len()" - same function returns the actual value of the param

File: test/gtest/uct/uct_test.cc [hunk 9]
```diff
@@ -213,7 +207,9 @@ void uct_test::flush() const {
                 ASSERT_UCS_OK(status);
             }
         }
-    } while (!flushed);
+    } while (!flushed && (ucs_get_time() < deadline));
+
+    return flushed ? UCS_OK : UCS_ERR_TIMED_OUT;
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: maybe better add EXPECT_xx here, instead of adding return value?

File: test/gtest/uct/uct_test.h [hunk 10]
```diff
@@ -189,15 +189,15 @@ class uct_test : public testing::TestWithParam<const resource*>,
     virtual void init();
     virtual void cleanup();
     virtual void modify_config(const std::string& name, const std::string& value);
-    virtual void get_config(const std::string& name, std::string& value);
+    virtual bool get_config(const std::string& name, std::string& value);
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: i think would be better to throw exception (e.g UCS_TEST_FAIL) rather than returning a value
(as general notice, the C++ way of handling error is to throw , and not return error code)

File: src/uct/ib/rc/base/rc_ep.h (hunk 1 above)
Comment (evgeny-leksikov) [inline src/uct/ib/rc/base/rc_ep.h]: As i understand you are moving condition from this place to another one `uct_rc_txqp_posted` because we will need to update `iface->tx.real_moderation` on every `iface->tx.cq_available` change, i.e. on every post send according to changes  in `iface->tx.cq_available` (i mean reservation of CQ credit on every post for error case) + on completion

File: src/uct/ib/rc/base/rc_ep.h (hunk 3 above)
Comment (evgeny-leksikov) [inline src/uct/ib/rc/base/rc_ep.h]: I do not think so, because then we can post 63 sends per connection without credits

File: src/uct/ib/rc/verbs/rc_verbs_iface.c (hunk 4 above)
Comment (evgeny-leksikov) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c]: To return TX_MODERATION credits

File: test/gtest/uct/test_peer_failure.cc (hunk 5 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: what do you mean? this is popping all elems from queue, not iterable container...
> Queue does not allow iteration through its elements

http://www.sgi.com/tech/stl/queue.html
any way if replace queue with vector, this becomes not relevant

File: test/gtest/uct/test_peer_failure.cc (hunk 6 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: Ok

File: test/gtest/uct/test_peer_failure.cc (hunk 7 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: I'm not sure about do this one-by-one, but may be to limit the same connection number, lets say from 2 to 10?

File: test/gtest/uct/test_peer_failure.cc (hunk 8 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: Ok

File: test/gtest/uct/uct_test.cc (hunk 9 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/uct_test.cc]: Ok

File: test/gtest/uct/uct_test.h (hunk 10 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/uct_test.h]: Not in this case... lack of value in the config should not fail the test and i do not think that catch an exception for fallback branch is a good usage of throw/catch mechanism 
(In general, this is a big debate about return code vs exceptions :) )

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4127/ for details (Mellanox internal link).

File: test/gtest/uct/test_peer_failure.cc (hunk 6 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: Test infrastructure is not ready for this

File: test/gtest/uct/test_peer_failure.cc (hunk 7 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: Test infrastructure is not ready for this

File: src/uct/ib/rc/base/rc_ep.h (hunk 1 above)
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: i see, need to keep thinking how we can remove the extra overhead from fast path

File: test/gtest/uct/test_peer_failure.cc (hunk 5 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_peer_failure.cc]: No way for queue, replaced with vector

File: src/uct/ib/rc/base/rc_ep.h (hunk 1 above)
Comment (evgeny-leksikov) [inline src/uct/ib/rc/base/rc_ep.h]: how about to set `txqp->unsignaled` to `inf` in `uct_rc_txqp_posted`?
`if (ucs_unlikely(iface->tx.cq_available -= res_count) <= 0)) {
    txqp->unsignaled_real = txqp->unsignaled;
    txqp->unsignaled = UINT16_MAX;
}`
this check in the same scope on cached value should be pretty fast,  also we need to check this value on completion to replace `inf` value of with saved real value (in the same cache line) with `unsignaled`  in "bad case"

On the next iteration post will be definitely "signaled" and the value will be initialized with zero as usual.

**Comment by evgeny-leksikov** [comment]
bot:bgate:retest
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4147/ for details (Mellanox internal link).

File: src/uct/ib/rc/base/rc_ep.h (hunk 1 above)
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.h]: @evgeny-leksikov sounds good, we need to also put the check for cq credits under that `if' to avoid overhead:
```
if (unlikely(cq_available < MAX_RES_COUNT)) {
   if (cq_available == 0) { 
       return UCS_ERR_NO_RESOURCE;  // use UCT_RC_CHECK_CQE macro
   };
   txqp->unsignaled_real = txqp->unsignaled; 
   txqp->unsignaled = UINT16_MAX;
}
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4148/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4164/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4168/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4176/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4181/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@MattBBaker pls share ORNL failure

**Comment by MattBBaker** [comment]
This error seems interesting, this should not be running out of memory.
```
[----------] 3 tests from rc/test_uct_pending
[ RUN      ] rc/test_uct_pending.pending_fairness/0
[pod9:24326:0] rc_verbs_ep.c:34   Fatal: ibv_post_send() returned 12 (Cannot allocate memory)
[pod9:24326:0] Process frozen...
make: *** [test] Terminated
```

**Comment by yosefe** [comment]
it indicates overflow in send queue

**Comment by evgeny-leksikov** [comment]
Thanks! the same as bgate, i'm debugging this.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4190/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4193/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4192/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4197/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4198/ for details (Mellanox internal link).

File: test/gtest/uct/test_peer_failure.cc [hunk 11]
```diff
@@ -314,10 +365,112 @@ UCS_TEST_P(test_uct_peer_failure_cb, desproy_ep_cb)
...
+void test_uct_peer_failure_multiple::init()
+{
+    if (RUNNING_ON_VALGRIND) {
+        /* See https://bugs.kde.org/show_bug.cgi?id=352742 */
+        UCS_TEST_SKIP_R("skipping on valgrind because \"brk segment overflow\"");
+    }
+
+    size_t tx_queue_len = get_tx_queue_len();
+
+#if defined(__arm__) || defined(__aarch64__)
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc]: do we still need that?
better check that page_size()>4k because this is relevant also for ppc64

**Comment by yosefe** [comment]
@evgeny-leksikov overall it looks good, can u pls fix the comment in the test and squash?
also, can you pls run osu_mbw_mr and osu_latency to make sure this patch does not affect performance? then i will +1
I also want to think if it's possible to implement it in a simpler way (e.g avoid unsignaled_store and unsignaled_store_count) but it's not a blocker if perf numbers are ok

**Comment by evgeny-leksikov** [comment]
@yosefe latency is not affected, osu_mbw_mr looks even a bit better vs master

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4205/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4207/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4208/ for details (Mellanox internal link).
